### PR TITLE
Add support for 2-d tensors to index_select and index_add.

### LINKED ISSUE #307: k2.index
I am thinking at the Python level we could add a function k2.index1 that does 1-d indexing with int32_t's, with special treatment of -1.  So like what PyTorch does with a[b], except it treats -1 separately and returns 0 in that case.

At the python level it could be something like the following:
```
   def index1(a, b):
     """ Return a[b] which is the elements of a indexed by tensor b.  
         Args:
           a: a torch.Tensor, of a fundamental type such as int32, int64, float, double.
               (We can later extend this to deal with other types as needed).  The number
               of axes should be 1 or 2 currently; and if 2, then a.stride[1] should equal the
               size of the type.
           b: a torch.Tensor of type torch.int32, and one axis.  Its elements must be
               either 0 <= i < a.shape[0], or -1.  If -1, then the corresponding element of
              the result will be 0. 
          Return:
              Returns a torch.Tensor whose first axis has dimension b.shape[0], and whose
              remaining axes (if any) are the same as a. 
      """
    pass
```
However the internal implementation would have to be done with a torch.Function, so we can implement the backward pass.

### PR DESCRIPTION
**Pull request by csukuangfj**
TODOs
- [x] Handle non-contiguous cases
- [x] add 2-d support to index_add

Closes #307

### CODE REVIEW

**Comment by danpovey** [comment]
OK with me. 
Note: there is a function IndexRows() in array_ops.h that indexes Array2 with Array1, but I'm not saying you should necessarily use it here.

File: k2/python/k2/autograd.py [hunk 1]
```diff
@@ -322,11 +322,14 @@ def forward(ctx, src: torch.Tensor, index: torch.Tensor) -> torch.Tensor:
...
         src, index = ctx.saved_tensors
 
-        ans = torch.zeros(src.size(0),
+        ans = torch.zeros(src.shape,
                           dtype=torch.float32,
                           device=src.device,
                           requires_grad=False)
-        _k2.index_add(index, out_grad, ans)
+        if out_grad.stride()[0] == 0:
+            # TODO(fangjun): why is the stride of out_grad 0 ??
```
Comment (csukuangfj) [inline k2/python/k2/autograd.py:330]: I tried to figure out why `out_grad.stride() == (0, 0)` but failed.
@danpovey @qindazhu could you give some comments about it?

File: k2/python/csrc/torch/index_add.cu [hunk 2]
```diff
@@ -52,68 +57,149 @@ void IndexAdd(torch::Tensor index, torch::Tensor value, torch::Tensor *in_out) {
...
+    K2_EVAL(
+        context, index_numel, lambda_contiguous, (int32_t i)->void {
+          int32_t in_out_i = index_data[i];
+          if (in_out_i == -1) return;
+
+          K2_DCHECK_GE(in_out_i, 0);
+          K2_DCHECK_LT(in_out_i, in_out_num_rows);
+          float *cur_in_out_data = in_out_data + in_out_i * in_out_num_cols;
+          const float *cur_value_data = value_data + i * in_out_num_cols;
+          for (int32_t j = 0; j != in_out_num_cols; ++j) {
```
Comment (qindazhu) [inline k2/python/csrc/torch/index_add.cu:143]: seems there's a loop in kernel, can't we add `j` in the parameter of lambda and use `Eval2`? BTW, I think we need also add a `TODO` here as we definitely need to find a way to avoid `atomicAdd` in future.

File: k2/python/csrc/torch/index_select.cu [hunk 3]
```diff
@@ -45,70 +50,116 @@ static torch::Tensor IndexSelect1D(torch::Tensor src, torch::Tensor index) {
...
+  if (src.is_contiguous() && index.is_contiguous()) {
+    K2_EVAL(
+        context, index_numel, lambda, (int32_t i)->void {
+          int32_t src_i = index_data[i];
+          const T *src_cur_row = src_data + src_i * src_num_cols;
+          T *ans_cur_row = ans_data + i * src_num_cols;
+          if (src_i != -1) {
+            K2_DCHECK_GE(src_i, 0);
+            K2_DCHECK_LT(src_i, src_num_rows);
+            for (int32_t j = 0; j != src_num_cols; ++j)
```
Comment (qindazhu) [inline k2/python/csrc/torch/index_select.cu:125]: add also move `j` here to the parameter of lambda and use Eval2

File: k2/python/csrc/torch/index_add.cu (hunk 2 above)
Comment (csukuangfj) [reply k2/python/csrc/torch/index_add.cu:143]: The number of loops is determined by the number of entries in the arc map, which is a 1-D tensor, so `Eval` is used here. 

I didn't find a way to avoid the loop here.

Comment (qindazhu) [reply k2/python/csrc/torch/index_add.cu:143]: we can just do like below
```
 K2_EVAL2(
        context, index_numel, *in_out_num_cols*,  lambda_contiguous, (int32_t i, *int32_t j*)->void {
          int32_t in_out_i = index_data[i];
          if (in_out_i == -1) return;

          K2_DCHECK_GE(in_out_i, 0);
          K2_DCHECK_LT(in_out_i, in_out_num_rows);
          float *cur_in_out_data = in_out_data + in_out_i * in_out_num_cols;
          const float *cur_value_data = value_data + i * in_out_num_cols;
#ifdef __CUDA_ARCH__
            atomicAdd(cur_in_out_data + j, cur_value_data[j]);
#else
            // for host thread, we assume single thread at present
            cur_in_out_data[j] += cur_value_data[j];
#endif
```

File: k2/python/k2/autograd.py (hunk 1 above)
Comment (qindazhu) [reply k2/python/k2/autograd.py:330]: have you ever printed some logs at C++ level of `index_select`?

Comment (csukuangfj) [reply k2/python/k2/autograd.py:330]: Yes, I have a check inside c++ to ensure that the stride for the column is 1. But the actual value is 0. That is
how I discovered that the `out_grad.stride() == (0, 0)`.

`out_grad` is generated by autograd from Python.

Comment (danpovey) [reply k2/python/k2/autograd.py:330]: What is the dimension of out_grad?

File: k2/python/csrc/torch/index_add.cu (hunk 2 above)
Comment (danpovey) [reply k2/python/csrc/torch/index_add.cu:143]: We have a __host__ version of atomicAdd so you don't need the ifdef.

**Comment by danpovey** [comment]
Guys, can you get out of the habit of doing force-push?  It makes it hard to view commits.
Better to merge and not rebase while you are working on a branch; I can do 'squash and merge' when merging from github.

File: k2/python/csrc/torch/index_add.cu [hunk 4]
```diff
@@ -22,12 +24,15 @@ namespace k2 {
...
-  // We support only 1-D tensors at present
+static void IndexAdd1D(torch::Tensor index, torch::Tensor value,
+                       torch::Tensor *in_out) {
+  NVTX_RANGE(K2_FUNC);
   K2_CHECK_EQ(index.scalar_type(), ToScalarType<int32_t>::value);
   K2_CHECK_EQ(index.dim(), 1);
 
-  K2_CHECK_EQ(value.scalar_type(), ToScalarType<float>::value);
+  K2_CHECK_EQ(value.scalar_type(), ToScalarType<float>::value)
+      << "index_add is supposed to be used in back propagation, "
```
Comment (danpovey) [inline k2/python/csrc/torch/index_add.cu:34]: I'd rather just say something to the effect that only float is supported right now.  It's a general purpose function.

File: k2/python/csrc/torch/index_add.cu [hunk 5]
```diff
@@ -52,68 +57,149 @@ void IndexAdd(torch::Tensor index, torch::Tensor value, torch::Tensor *in_out) {
...
+        atomicAdd(in_out_data + in_out_i * in_out_stride,
+                  value_data[i * value_stride]);
+#else
+      // for host thread, we assume single thread at present
+      in_out_data[in_out_i * in_out_stride] += value_data[i * value_stride];
+#endif
+      });
+}
+
+static void IndexAdd2D(torch::Tensor index, torch::Tensor value,
```
Comment (danpovey) [inline k2/python/csrc/torch/index_add.cu:101]: There is no documentation here!!
You should be writing the documentation first!

Comment (danpovey) [reply k2/python/csrc/torch/index_add.cu:101]: I can't review this without documentation- how can I tell if it is doing what is intended when that is not stated?

**Comment by danpovey** [comment]
The structure of this is problematic.  It's in torch/.  Remember, the code in torch/ is intended to be a fairly thin wrapper for our library.  The actual functionality should probably be implemented in tensor_ops.cu, and declared in tensor_ops.h.

File: k2/python/k2/autograd.py (hunk 1 above)
Comment (qindazhu) [reply k2/python/k2/autograd.py:330]: > I have a check inside c++ to ensure that the stride for the column is 1. But the actual value is 0

I suppose you add a check locally(sorry, I didn't find it with current code), so you mean even in C++, the stride is 0? when we create tensor with something like `src.options()`? what if we add `stride` manually?

**Comment by danpovey** [comment]
If the dimension is 1 it might be OK that the stride is 0.  In that case we
could treat the stride as being whatever value we like.

**Comment by danpovey** [comment]
... in fact, I have already declared functions IndexSelect() and IndexAdd() in tensor.h, just not implemented them yet.
The idea is that *where possible* these functions will call corresponding functions declared for Array1 and Array2.
There may be cases where, due to strides etc., this is not possible.

Note the functgions Index() and IndexRows() in array_ops.h.

We don't want the same code implemented in different parts of the software-- better to have one place where each operation is implemented, for easier optimization!

**Comment by csukuangfj** [comment]
> If the dimension is 1 it might be OK that the stride is 0.  In that case we
could treat the stride as being whatever value we like.

The problem is that `out_grad` is a 2-d tensor and `out_grad.stride() == (0, 0)`, 
whereas `out_grad.clone().stride() == (some_nonzero_value, 1)`. I cannot figure out the reason.

----

> Guys, can you get out of the habit of doing force-push?

Sorry, no force-push from now on.

---

> The structure of this is problematic.

Will refactor it as suggested.

**Comment by danpovey** [comment]
Thanks!!

**Comment by danpovey** [comment]
So was the shape of `out_grad` in this case (1, 1)?  In that case we can ignore the stride, or treat it as whatever we want.

**Comment by danpovey** [comment]
So was the shape of `out_grad` in this case (1, 1)?  In that case we can ignore the stride, or treat it as whatever we want.

**Comment by csukuangfj** [comment]
> So was the shape of out_grad in this case (1, 1)

The `out_grad.shape` contains two very large positive numbers that are generated by `torch.randint`.
It's strange that its stride is `(0, 0)`. I have never encountered this problem before.

**Comment by danpovey** [comment]
I wonder whether torch could be doing something where it lazily allocates
things and doesn't decide the stride unless you attempt to access the data?
When I was designing a tensor toolkit previously I was planning to do
something like that
(although in that case the strides would have been defined).

File: k2/python/k2/autograd.py (hunk 1 above)
Comment (csukuangfj) [reply k2/python/k2/autograd.py:330]: > I suppose you add a check locally(sorry, I didn't find it with current code),

Here is the check in the code:

https://github.com/k2-fsa/k2/blob/da3c30e0d277a69013610bf0838592f25e003861/k2/python/csrc/torch/index_add.cu#L167

where `value` is actually `out_grad` from Python:
https://github.com/k2-fsa/k2/blob/da3c30e0d277a69013610bf0838592f25e003861/k2/python/k2/autograd.py#L329-L332

**Comment by danpovey** [comment]
I see now why the stride is zero.
If Torch was backpropping the result of you summing all the elements of a tensor, then the gradient would be the same for all elements, so the easiest way for Torch to do the backprop for that would be take a scalar Tensor containing 1 and just view it as having shape (x,y), with zero stride.
It's actually quite nice that it does that.

File: k2/python/k2/autograd.py (hunk 1 above)
Comment (csukuangfj) [reply k2/python/k2/autograd.py]: fixed.

File: k2/python/csrc/torch/index_add.cu (hunk 2 above)
Comment (csukuangfj) [reply k2/python/csrc/torch/index_add.cu]: Fixed.

There was no `__host__ `version of atomicAdd.  A host version is added in this pullrequest.

File: k2/python/csrc/torch/index_select.cu (hunk 3 above)
Comment (csukuangfj) [reply k2/python/csrc/torch/index_select.cu]: fixed.

File: k2/python/csrc/torch/index_add.cu (hunk 4 above)
Comment (csukuangfj) [reply k2/python/csrc/torch/index_add.cu]: Fixed by removal.

File: k2/python/csrc/torch/index_add.cu (hunk 5 above)
Comment (csukuangfj) [reply k2/python/csrc/torch/index_add.cu]: Removed.

**Comment by csukuangfj** [comment]
Ready for review.

**Comment by danpovey** [comment]
OK, I have one point of discussion...

I notice you are implementing the operations directly in class Tensor.
It would be possible, if the strides are suitable, to use the existing implementations involving Array1 and Array2,
e.g. Index(), IndexRows().  [We can consider changing these names for consistency.]

The advantage of using the existing implementations is (1) slight kernel simplification due to
knowing some of the strides are 1, and (2) Having a smaller number of locations to optimize.

I may be overthinking this though.  We could perhaps merge as-is but just make a note for the future to address this later.

**Comment by csukuangfj** [comment]
> It would be possible, if the strides are suitable, to use the existing implementations involving Array1 and Array2,
e.g. Index(), IndexRows()

Yes, I am using Index() and IndexRows() from array_ops when the stride is suitable:
https://github.com/k2-fsa/k2/blob/fc9e7154a11f3cd3a4d3ab953f37c20423287ee6/k2/python/csrc/torch/index_select.cu#L55-L60

https://github.com/k2-fsa/k2/blob/fc9e7154a11f3cd3a4d3ab953f37c20423287ee6/k2/python/csrc/torch/index_select.cu#L95-L99

There is no `IndexAdd` in array_ops but there is `IndexAdd` declared in tensor_ops, so I
implemented IndexAdd in tensor_ops and wrapped it to Python.

**Comment by danpovey** [comment]
OK cool, thanks!  Merging.

**Comment by qindazhu** [comment]
Seems there's a build failure? 
```
/home/runner/work/k2/k2/k2/csrc/utils.h(518): error: no instance of overloaded function "atomicAdd" matches the argument list
            argument types are: (double *, double)
          detected during:
            instantiation of "void k2::AtomicAdd(T *, T) [with T=double]" 
/home/runner/work/k2/k2/k2/csrc/tensor_ops.cu(298): here
            instantiation of "void k2::IndexAdd1DImpl(k2::ContextPtr, const T *, int32_t, int32_t, const int32_t *, __nv_bool, int32_t, int32_t, T *) [with T=double]" 
/home/runner/work/k2/k2/k2/csrc/tensor_ops.cu(373): here
```

**Comment by csukuangfj** [comment]
> Seems there's a build failure?


Fixed in #454

**Comment by danpovey** [comment]
Oops.  That may relate to different CUDA versions, possibly?  Looks like atomicAdd may only be implemented from CUDA version 6.0.

**Comment by csukuangfj** [comment]
> there is a way, it seems, to implement it for earlier CUDA versions. We can maybe have an #ifdef on the number of __CUDA_ARCH__ ?

I just removed `double`.

**Comment by danpovey** [comment]
OK I can merge your change for now to fix the build problem, but can you
please either make an issue or make a PR to fix it differently?

**Comment by csukuangfj** [comment]
> OK I can merge your change for now to fix the build problem, but can you
please either make an issue or make a PR to fix it differently?

Fixed in #455
